### PR TITLE
Add new consul modules and reuse code between them.

### PR DESCRIPTION
**Pull request by apollo13**

### CODE REVIEW

File: plugins/modules/consul_acl.py [hunk 1]
```diff
@@ -26,6 +26,10 @@
     support: none
   diff_mode:
     support: none
+deprecated:
+  removed_in: 99.0.0 # TODO
```
Comment (felixfontein) [inline plugins/modules/consul_acl.py:30]:
```suggestion
  removed_in: 10.0.0
```

File: plugins/modules/consul_auth_method.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+module: consul_auth_method
+short_description: Manipulate Consul auth methods
+version_added: 0.0.0
```
Comment (felixfontein) [inline plugins/modules/consul_auth_method.py:15]:
```suggestion
version_added: 8.3.0
```

File: plugins/modules/consul_policy.py [hunk 3]
```diff
@@ -20,12 +21,13 @@
...
   - community.general.consul
+  - community.general.consul.token
   - community.general.attributes
 attributes:
   check_mode:
-    support: none
+    support: full
   diff_mode:
-    support: none
+    support: partial
```
Comment (felixfontein) [inline plugins/modules/consul_policy.py:30]: If this is new, it needs a changelog fragment, and
```suggestion
    support: partial
    version_added: 8.3.0
```
(same for check mode)

Comment (felixfontein) [reply plugins/modules/consul_policy.py:30]: Also, `support: partial` should come with `details:` which accepts a list of strings (paragraphs) explaining why it's not `support: full`.

File: meta/runtime.yml [hunk 4]
```diff
@@ -22,6 +22,10 @@ plugin_routing:
     nios_next_network:
       redirect: infoblox.nios_modules.nios_next_network
   modules:
+    consul_acl:
```
Comment (felixfontein) [inline meta/runtime.yml:25]: It's best to move the deprecation to a separate PR. It also needs a separate changelog fragment (`deprecated_features`).

**Review by felixfontein — COMMENTED** [review]
Some first comments :)

File: meta/runtime.yml (hunk 4 above)
Comment (apollo13) [reply meta/runtime.yml]: Ok removed it for now from this PR completely.

**Comment by apollo13** [comment]
I think the 3rd party tests are failing due to:
> filter community.general.time missing documentation (or could not parse documentation): No documentation available for community.general.time (/tmp/tmprzds3cf4/ansible_collections/community/general/plugins/filter/time.py)\

which I didn't touch.

**Comment by felixfontein** [comment]
The Zuul test is always failing, you can ignore it. I'm currently trying to get completely rid of it, since it's very slow and fails relatively often (and for some time now constantly fails).

**Comment by felixfontein** [comment]
Can you modify changelogs/fragments/7826-consul-modules-refactoring.yaml by adding this PR to the brackets at the end (in format `... (url1, url2).`)?

File: plugins/modules/consul_auth_method.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      - Free form human readable description of the auth method.
+    type: str
+  display_name:
+    description:
+      - An optional name to use instead of O(name) when displaying information about this auth method.
+    type: str
+  max_token_ttl:
+    description:
+      - This specifies the maximum life of any token created by this auth method.
+      - Can be specified in the form of "60s" or "5m" (i.e., 60 seconds or 5 minutes, respectively).
```
Comment (felixfontein) [inline plugins/modules/consul_auth_method.py:64]:
```suggestion
      - Can be specified in the form of V(60s) or V(5m) (that is, 60 seconds or 5 minutes, respectively).
```

File: plugins/modules/consul_policy.py [hunk 6]
```diff
@@ -20,12 +21,15 @@
   - Håkon Lerring (@Hakon)
 extends_documentation_fragment:
   - community.general.consul
+  - community.general.consul.token
   - community.general.attributes
 attributes:
   check_mode:
-    support: none
+    support: full
```
Comment (felixfontein) [inline plugins/modules/consul_policy.py:28]: Here as well:
```suggestion
    support: full
    version_added: 8.3.0
```

File: plugins/modules/consul_policy.py [hunk 7]
```diff
@@ -95,155 +97,64 @@
 """
 
 RETURN = """
+policy:
+    description: The policy as returned by the consul HTTP API
```
Comment (felixfontein) [inline plugins/modules/consul_policy.py:101]:
```suggestion
    description: The policy as returned by the consul HTTP API.
```

File: plugins/modules/consul_role.py [hunk 8]
```diff
@@ -20,12 +21,15 @@
...
+  - community.general.consul.token
   - community.general.attributes
 attributes:
   check_mode:
     support: full
   diff_mode:
-    support: none
+    support: partial
+    details:
+      - In check mode the diff will miss operational attributes.
```
Comment (felixfontein) [inline plugins/modules/consul_role.py:32]:
```suggestion
      - In check mode the diff will miss operational attributes.
    version_added: 8.3.0
```

File: plugins/modules/consul_role.py [hunk 9]
```diff
@@ -61,6 +65,22 @@
           - The ID of the policy to attach to this role; see M(community.general.consul_policy) for more info.
           - Either this or O(policies[].name) must be specified.
         type: str
+  templated_policies:
+    description:
+      - The list of templated policies that should be applied to the role.
+    type: list
+    elements: dict
```
Comment (felixfontein) [inline plugins/modules/consul_role.py:72]:
```suggestion
    elements: dict
    version_added: 8.3.0
```

File: plugins/modules/consul_role.py [hunk 10]
```diff
@@ -69,13 +89,15 @@
       - If not specified, any service identities currently assigned will not be changed.
       - If the parameter is an empty array (V([])), any node identities assigned will be unassigned.
     suboptions:
-      name:
+      service_name:
         description:
           - The name of the node.
           - Must not be longer than 256 characters, must start and end with a lowercase alphanumeric character.
           - May only contain lowercase alphanumeric characters as well as - and _.
```
Comment (felixfontein) [inline plugins/modules/consul_role.py:96]:
```suggestion
          - May only contain lowercase alphanumeric characters as well as - and _.
          - This suboption has been renamed from O(service_identities[].name) to O(service_identities[].service_name)
            in community.general 8.3.0. The old name can still be used.
```

File: plugins/modules/consul_role.py [hunk 11]
```diff
@@ -93,20 +114,22 @@
       - If not specified, any node identities currently assigned will not be changed.
       - If the parameter is an empty array (V([])), any node identities assigned will be unassigned.
     suboptions:
-      name:
+      node_name:
         description:
           - The name of the node.
           - Must not be longer than 256 characters, must start and end with a lowercase alphanumeric character.
           - May only contain lowercase alphanumeric characters as well as - and _.
```
Comment (felixfontein) [inline plugins/modules/consul_role.py:121]: Same here.

File: plugins/modules/consul_token.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    description:
+      - List of service identities to attach to the token.
+      - If not specified, any service identities currently assigned will not be changed.
+      - If the parameter is an empty array (V([])), any node identities assigned will be unassigned.
+    suboptions:
+      service_name:
+        description:
+          - The name of the service.
+          - Must not be longer than 256 characters, must start and end with a lowercase alphanumeric character.
+          - May only contain lowercase alphanumeric characters as well as - and _.
```
Comment (felixfontein) [inline plugins/modules/consul_token.py:118]:
```suggestion
          - May only contain lowercase alphanumeric characters as well as V(-) and V(_).
```

File: plugins/modules/consul_token.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    description:
+      - List of node identities to attach to the token.
+      - If not specified, any node identities currently assigned will not be changed.
+      - If the parameter is an empty array (V([])), any node identities assigned will be unassigned.
+    suboptions:
+      node_name:
+        description:
+          - The name of the node.
+          - Must not be longer than 256 characters, must start and end with a lowercase alphanumeric character.
+          - May only contain lowercase alphanumeric characters as well as - and _.
```
Comment (felixfontein) [inline plugins/modules/consul_token.py:140]:
```suggestion
          - May only contain lowercase alphanumeric characters as well as V(-) and V(_).
```

File: plugins/modules/consul_token.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+        required: true
+  local:
+    description:
+      - If true, indicates that the token should not be replicated globally
+        and instead be local to the current datacenter.
+    type: bool
+  expiration_ttl:
+    description:
+      - This is a convenience field and if set will initialize the C(expiration_time).
+        Can be specified in the form of "60s" or "5m" (i.e., 60 seconds or 5 minutes,
```
Comment (felixfontein) [inline plugins/modules/consul_token.py:157]:
```suggestion
        Can be specified in the form of V(60s) or V(5m) (that is, 60 seconds or 5 minutes,
```

File: plugins/modules/consul_token.py [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+- name: Delete a token
+  community.general.consul_token:
+    state: absent
+    accessor_id: 07a7de84-c9c7-448a-99cc-beaf682efd21
+    token: 8adddd91-0bd6-d41d-ae1a-3b49cfa9a0e8
+"""
+
+RETURN = """
+token:
+    description: The token as returned by the consul HTTP API
```
Comment (felixfontein) [inline plugins/modules/consul_token.py:188]:
```suggestion
    description: The token as returned by the consul HTTP API.
```

File: plugins/modules/consul_acl_bootstrap.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+EXAMPLES = """
+- community.general.consul_acl_bootstrap:
+    bootstrap_secret: 22eaeed1-bdbd-4651-724e-42ae6c43e387
+"""
+
+RETURN = """
+result:
+    description:
+      - The bootstrap result as returned by the consul HTTP API
```
Comment (felixfontein) [inline plugins/modules/consul_acl_bootstrap.py:51]:
```suggestion
      - The bootstrap result as returned by the consul HTTP API.
```

File: plugins/modules/consul_acl_bootstrap.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+options:
+  state:
+    description:
+      - Whether the token should be present or absent.
+    choices: ['present', 'bootstrapped']
+    default: present
+    type: str
+  bootstrap_secret:
+    description:
+      - The secret to be used as secret id for the initial token.
```
Comment (felixfontein) [inline plugins/modules/consul_acl_bootstrap.py:38]:
```suggestion
      - The secret to be used as secret ID for the initial token.
```

File: plugins/modules/consul_auth_method.py [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+- name: delete auth method
+  community.general.consul_auth_method:
+    name: test
+    state: absent
+    token: "{{ consul_management_token }}"
+"""
+
+RETURN = """
+auth_method:
+  description: The auth method as returned by the consul HTTP API
```
Comment (felixfontein) [inline plugins/modules/consul_auth_method.py:107]:
```suggestion
  description: The auth method as returned by the consul HTTP API.
```

File: plugins/modules/consul_binding_rule.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+- name: remove a binding rule
+  community.general.consul_binding_rule:
+    name: my_name
+    auth_method: minikube
+    state: absent
+"""
+
+RETURN = """
+binding_rule:
+  description: The binding rule as returned by the consul HTTP API
```
Comment (felixfontein) [inline plugins/modules/consul_binding_rule.py:95]:
```suggestion
  description: The binding rule as returned by the consul HTTP API.
```

**Review by felixfontein — COMMENTED** [review]
For new features (check/diff mode support, new options) and option renames you need a changelog fragment (`minor_changes`).

**Comment by apollo13** [comment]
> For new features (check/diff mode support, new options) and option renames you need a changelog fragment (`minor_changes`).

Can I roll that into the existing fragment, or does it need a new one?

File: plugins/modules/consul_acl_bootstrap.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+    type: str
+  bootstrap_secret:
+    description:
+      - The secret to be used as secret ID for the initial token.
+      - Needs to be an UUID.
+    type: str
+"""
+
+EXAMPLES = """
+- community.general.consul_acl_bootstrap:
```
Comment (felixfontein) [inline plugins/modules/consul_acl_bootstrap.py:44]:
```suggestion
- name: ...
  community.general.consul_acl_bootstrap:
```

File: plugins/modules/consul_auth_method.py [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+  config:
+    description:
+      - The raw configuration to use for the chosen auth method.
+      - Contents will vary depending upon the type chosen.
+      - Required when the auth method is created.
+    type: dict
+"""
+
+EXAMPLES = """
+- name: create an auth method
```
Comment (felixfontein) [inline plugins/modules/consul_auth_method.py:80]: Task names should always start with an uppercase letter:
```suggestion
- name: Create an auth method
```

File: plugins/modules/consul_auth_method.py [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+          4lgOEePzNm0tRgeLezV6ffAt0gunVTLw7onLRnrq0/IzW7yWR7QkrmBL7jTKEn5u
+          +qKhbwKfBstIs+bMY2Zkp18gnTxKLxoS2tFczGkPLPgizskuemMghRniWaoLcyeh
+          kd3qqGElvW/VDL5AaWTg0nLVkjRo9z+40RQzuVaE8AkAFmxZzow3x+VJYKdjykkJ
+          0iT9wCS0DRTXu269V264Vf/3jvredZiKRkgwlL9xNAwxXFg0x/XFw005UWVRIkdg
+          cKWTjpBP2dPwVZ4WWC+9aGVd+Gyn1o0CLelf4rEjGoXbAAEgAqeGUxrcIlbjXfbc
+          mwIDAQAB
+          -----END PUBLIC KEY-----
+    token: "{{ consul_management_token }}"
+
+- name: delete auth method
```
Comment (felixfontein) [inline plugins/modules/consul_auth_method.py:98]:
```suggestion
- name: Delete auth method
```

File: plugins/modules/consul_binding_rule.py [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+      - What it binds to can be adjusted with different values of the O(bind_type) parameter.
+    type: str
+  bind_vars:
+    description:
+      - Specifies the templated policy variables when O(bind_type) is set to V(templated-policy).
+    type: dict
+"""
+
+EXAMPLES = """
+- name: create a binding rule
```
Comment (felixfontein) [inline plugins/modules/consul_binding_rule.py:77]:
```suggestion
- name: Create a binding rule
```

File: plugins/modules/consul_binding_rule.py [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+  community.general.consul_binding_rule:
+    name: my_name
+    description: example rule
+    auth_method: minikube
+    bind_type: service
+    bind_name: "{{ serviceaccount.name }}"
+    token: "{{ consul_management_token }}"
+
+- name: remove a binding rule
```
Comment (felixfontein) [inline plugins/modules/consul_binding_rule.py:86]:
```suggestion
- name: Remove a binding rule
```

**Comment by felixfontein** [comment]
@apollo13 thanks a lot for your contributions!
